File: javaexps/exp002/src/util/NamedSet.java
package util;

import java.util.*;

public class NamedSet<T extends NamedObject> implements Iterable<T> {
    private final SortedMap<String, T> data;

    public NamedSet() {
        this.data = new TreeMap<>();
    }

    public boolean isEmpty() {
        return this.data.isEmpty();
    }

    public boolean contains(final T named) {
        if (named == null) return false;
        return this.data.containsKey(named.name());
    }

    public boolean contains(final String name) {
        return this.data.containsKey(name);
    }

    public Iterator<T> iterator() {
        return this.data.values().iterator();
    }

    public Iterator<String> names() {
        return this.data.keySet().iterator();
    }

    public T get(final String name) {
        return this.data.get(name);
    }

    public boolean add(final T named) {
        if (named == null) return false;
        final String name = named.name();
        if (this.data.containsKey(name)) return false;
        this.data.put(name, named);
        return true;
    }

    public boolean remove(final T named) {
        if (named == null) return false;
        final String name = named.name();
        T other = this.data.get(name);
        if (named != other) return false;
        return this.data.containsKey(name);
    }

    public boolean remove(final String name) {
        return this.data.containsKey(name);
    }
}

File: javaexps/exp002/src/named/NamedSet.kt
package named

import json.JsonMap
import json.JsonObj
import json.Jsonable
import kotlin.collections.Collection
import kotlin.collections.HashMap

class NamedSet<T: NamedObject>: Collection<T>, Jsonable {
    private val data = HashMap<String, T>()
    
    override val size: Int get() = this.data.size

    override fun isEmpty() = this.data.isEmpty()

    override fun containsAll(elements: Collection<T>) = elements.all { this.contains(it) }

    override fun contains(element: T) = this.data[element.name] == element

    fun contains(name: String) = this.data.containsKey(name)

    operator fun get(name: String) = this.data[name]

    override fun iterator() = this.data.values.iterator()

    fun add(element: T): Boolean {
        if (this.contains(element.name)) return false
        this.data[element.name] = element
        return true
    }

    fun remove(element: T): Boolean {
        if (!contains(element)) return false
        data.remove(element.name)
        return true
    }

    fun remove(name: String): Boolean {
        if (!contains(name)) return false
        data.remove(name)
        return true
    }

    override fun toJson(): JsonObj? {
        val m = JsonMap()
        for ((name, elem) in this.data)
            m[name] = elem.toJson()
        return m
    }
}